### PR TITLE
Add plain HTTP transport to OpAMP

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Resolves https://github.com/open-telemetry/opamp-spec/issues/50

Certain applications may not need the OpAMP's capabilities enabled by WebSocket (ability to push configs from server without Agent's polling) and may prefer to implement a simpler HTTP transport. This is likely the case for short-lived Agents such as for example the ones running in AWS lambda. The HTTP transport is likely simpler to implement on the Server (easier to manage connections) and also removes an additional dependency (a WebSocket library) from the client.

We are also considering implementing OpAMP in OpenTelemetry SDKs, which can result in significantly larger number of OpAMP Agents connecting to the Server, so the Server has to handle a larger number of persistent WebSocket connections, which may pose more burden than necessary for use cases where there are no benefits from having a persistent connection (e.g. the above mentioned short-lived use case).

This changes defines a plain HTTP transport for OpAMP. It is a fairly simple modification of the spec to allow it, with AgentToServer/ServerToAgent messages exchanged via HTTP POST Body.

It is also not very complicated to support both WebSocket and plain HTTP on the same Server and if we want to make it seamless from Agent's perspective we can add an indicator of the desired transport in the initial connection attempt made by the Agent, so that the Server knows whether to upgrade to WebSocket or just reply in plain HTTP. The "Content-Type" request header can be such indicator. We can add this in a follow-up PR.

Here is a draft Go OpAMP server implementation for reference: https://github.com/open-telemetry/opamp-go/compare/main...tigrannajaryan:plain-http-transport

### CODE REVIEW

**Comment by anuraggujral** [comment]
In the above PR, it is not clear from the second paragraph that here we are considering OpenTelemetry SDKs as agents, may be very should make it more explicit.

**Comment by tigrannajaryan** [comment]
@pmm-sumo what do you think about this change?

File: specification.md
Comment (pmm-sumo) [inline specification.md:175]: I think it would be worth considering allowing JSON serialized Protobuf as well. It typically has less dependencies and in some cases (e.g. mobile apps, web browsers, Node.js) is preferred over Protobuf

File: specification.md
Comment (pmm-sumo) [inline specification.md:186]: I think this is the right approach. Polling also acts as a heartbeat from the agent

File: specification.md
Comment (pmm-sumo) [inline specification.md:195]: It took me a while to process the last sentence so just wanted to make sure I got it right. When Agent has something to send to the Server, it does not need to wait for the polling interval to kick, it can send the message anytime, for which the server might return another message (and the conversation might continue further). 

When Server has something to send to the Agent (e.g. ConnectioSettingsOffer), it might need to wait for the "ping" from the Agent.

File: specification.md
Comment (pmm-sumo) [inline specification.md:201]: Should we add something about compression or perhaps we just assume it's a transport propety?

**Comment by pmm-sumo** [comment]
> @pmm-sumo what do you think about this change?

I think it will be a very useful addition. I think it might be worth running by OpenTelemetry SDK's maintainers, since this is one of the use cases for it

File: specification.md
Comment (tigrannajaryan) [reply specification.md:175]: Yes, that's fairly easily achievable via Content-Type header and also easy to implement on the server-side to support both JSON and binary Protobufs.

File: specification.md
Comment (tigrannajaryan) [inline specification.md:191]:
```suggestion
When using HTTP transport the sequence of messages is exactly the same as it is
```

File: specification.md
Comment (tigrannajaryan) [reply specification.md:195]: Correct. I probably need to rewrite this to make it clear.

File: specification.md
Comment (tigrannajaryan) [reply specification.md:201]: Good question. I think for the response compression we can leave it to be negotiated via Accept-Encoding/Content-Encoding headers.

There is no standard HTTP way to do the negotiation for the request. So, if we want the compression possibility either we mandate its support in the spec or leave it undefined (and maybe say that it is user configurable). For OTLP we chose to mandate "none" and "gzip" in the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#protocol-details
So maybe we should do the same here.

File: specification.md
Comment (tigrannajaryan) [reply specification.md:175]: Added as Open Question.

File: specification.md
Comment (tigrannajaryan) [reply specification.md:206]: Added to the text.

File: specification.md
Comment (tigrannajaryan) [reply specification.md]: Made some clarifications.

**Review by pmm-sumo — APPROVED** [review]

**Pull request merged**
